presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,265 @@
+import com.facebook.presto.sql.tree.SimpleGroupBy;
+import com.facebook.presto.sql.tree.SingleColumn;
+import com.facebook.presto.sql.tree.SortItem;
+import com.facebook.presto.sql.tree.Table;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.analyzer.MaterializedViewInformationExtractor.MaterializedViewInfo;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewQueryOptimizer
+        extends AstVisitor<Node, Void>
+{
+    private final Table materializedView;
+    private final Query materializedViewQuery;
+
+    private MaterializedViewInfo materializedViewInfo;
+
+    private static final Logger logger = Logger.get(MaterializedViewQueryOptimizer.class);
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query materializedViewQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+        this.materializedViewQuery = requireNonNull(materializedViewQuery, "materialized view query is null");
+    }
+
+    public Node rewrite(Node node)
+    {
+        try {
+            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+            materializedViewInformationExtractor.process(materializedViewQuery);
+            this.materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
+            return process(node);
+        }
+        catch (Exception ex) {
+            logger.error(ex, "Failed to rewrite query: %s.", node);
+            return node;
+        }
+    }
+
+    @Override
+    protected Node visitNode(Node node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitQuery(Query node, Void context)
+    {
+        QueryBody rewrittenQueryBody = (QueryBody) process(node.getQueryBody(), context);
+        return new Query(
+                node.getWith(),
+                rewrittenQueryBody,
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
+    {
+        if (!node.getFrom().isPresent()) {
+            throw new IllegalStateException("Query with no From clause is not rewritable by materialized view");
+        }
+        Optional<Relation> rewrittenFrom = node.getFrom().map(from -> (Relation) process(from, context));
+        Select rewrittenSelect = (Select) process(node.getSelect(), context);
+
+        // TODO: Handle filter containment problem https://github.com/prestodb/presto/issues/16405
+        if (materializedViewInfo.getWhereClause().isPresent() && !materializedViewInfo.getWhereClause().equals(node.getWhere())) {
+            throw new IllegalStateException("Query with no groupBy clause is not rewritable by materialized view with groupBy clause");
```

Comment:
Typo in the exception message: WhereClause not groupBy clause.

Reply:
We should be able to move this where cluase check int he visitFrom?

Reply:
@jainxrohit Similar to my argument in visitGroupBy, visitWhere is only processed if there exists a filter condition in the base query. This IF statement is a quick check to eliminate the "invalid" cases, which will be improved in future PR.